sda/cmd/rotatekey/rotatekey.go

```diff
@@ -0,0 +1,268 @@
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			// Fetch rotate key hash before starting work so that we make sure the hash state
+			// has not changed since the application startup.
+			keyhash := hex.EncodeToString(conf.RotateKey.PublicKey[:])
+			// exit app if target key was modified after app start-up, e.g. if key has been deprecated
+			if err = db.CheckKeyHash(keyhash); err != nil {
+				panic(fmt.Errorf("check of target key failed, reason: %v", err))
+			}
+
+			// we unmarshal the message in the validation step so this is safe to do
+			_ = json.Unmarshal(delivered.Body, &message)
+
+			fileID := message.FileID
+
+			// Get current keyhash for the file, send to error queue if this fails
+			oldKeyHash, err := db.GetKeyHash(fileID)
+			if err != nil {
+				msg := fmt.Sprintf("failed to get keyhash for file with file-id: %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			// Check that the file is not already encrypted with the target key
+			if oldKeyHash == keyhash {
+				log.Infof("the file with file-id: %s is already encrypted with the given rotation c4gh key", fileID)
+				if err := delivered.Ack(false); err != nil {
+					log.Errorf("failed to ack following already encrypted with key message")
+				}
+
+				continue
+			}
+
+			// reencrypt header
+			log.Debugf("rotating c4gh key for file with file-id: %s", fileID)
+
+			header, err := db.GetHeader(fileID)
+			if err != nil {
+				msg := fmt.Sprintf("GetHeader failed for file-id: %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			newHeader, err := reencrypt.CallReencryptHeader(header, pubKeyEncoded, conf.RotateKey.Grpc)
+			if err != nil {
+				msg := fmt.Sprintf("failed to rotate c4gh key for file %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+			if newHeader == nil {
+				err := errors.New("reencrypt returned empty header")
+				msg := fmt.Sprintf("failed to rotate c4gh key for file %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			// Rotate header and keyhash in database
+			if err := db.RotateHeaderKey(newHeader, keyhash, fileID); err != nil {
+				msg := fmt.Sprintf("RotateHeaderKey failed for file-id: %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
```

Comment:
This should result in a requeueing of the message.

Reply:
Not really.
An error in that DB function means that the header and key reference is unchanged and as a result the work needs to be performed again and thus the message should be requeued.